File: Olympiad/COCI/official/2007/contest5_solutions/dvaput.pas

(*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task DVAPUT
  Programming language Pascal
*)

program dvaput;

const
   MAXLEN  = 200000;
   MAXHASH = 200003;

type node = record
   hash, key, next : longint;
end;

var
   nnodes         : longint;
   nodes          : array[1..MAXLEN] of node;
   tablica        : array[0..MAXHASH-1] of longint;
   L              : longint;
   str            : array[1..MAXLEN] of char;
   i, lo, hi, mid : longint;

col, ins:longint;

function trazi2(K : longint) : boolean;
var
   hash, sub, i, p, j : longint;
   isti               : boolean;
begin
   if K = 0 then begin
      trazi2 := true;
      exit;
   end;

   while nnodes > 0 do begin
      tablica[ nodes[nnodes].hash ] := -1;
      nnodes := nnodes-1;
   end;

   hash := 0; sub := 1;
   for i:=1 to K do begin
      sub := (26*sub) mod MAXHASH;
      hash := (26*hash + ord(str[i])) mod MAXHASH;
   end;

   sub := MAXHASH - sub;

   for i:=K to L do begin
      p := tablica[hash];
      while p <> -1 do begin
         isti := true;
         for j:=0 to K-1 do
            if str[ nodes[p].key+j ] <> str[ i-K+1+j ] then begin
               isti := false;
               break;
            end;

         if isti then begin
            trazi2 := true;
            exit;
         end;

         p := nodes[p].next;
      end;

      ins := ins+1;
      if tablica[hash] <> -1 then
         col := col+1;
      nnodes := nnodes+1;
      nodes[nnodes].hash := hash;
      nodes[nnodes].key  := i-K+1;
      nodes[nnodes].next := tablica[hash];
      tablica[hash] := nnodes;

      hash := (26*hash + sub*ord(str[i-K+1]) + ord(str[i+1])) mod MAXHASH;
   end;

   trazi2 := false;
end;

begin
   readln(L);
   for i:=1 to L do
      read(str[i]);

   for i:=0 to MAXHASH-1 do
      tablica[i] := -1;

   lo := 0; hi := L-1;
   while lo<hi do begin
      mid := (lo+hi+1) div 2;

      if trazi2(mid) then
         lo := mid
      else
         hi := mid-1;
   end;

   writeln(lo);
end.

File: Olympiad/COCI/official/2007/contest5_solutions/dvaput.c

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 5 - Task DVAPUT
  Programming language C
*/

#include <stdio.h>
#include <string.h>

#define MAXLEN  200000
#define MAXHASH 200003

typedef struct {
   int hash, key, next;
} node;

int nnodes = 0;
node nodes[MAXLEN];
int tablica[MAXHASH];

int trazi2( const char *str, int len )
{
   unsigned hash, sub;
   int i;

   if ( len == 0 ) return 1;

   while ( nnodes > 0 ) {
      tablica[ nodes[--nnodes].hash ] = -1;
   }

   hash = 0; sub = 1;
   for ( i=0; i<len; ++i ) {
      sub = (26*sub) % MAXHASH;
      hash = (26*hash + str[i]) % MAXHASH;
   }

   sub = MAXHASH - sub;
   for ( i=len-1; str[i]!='\0'; ++i ) {
      int p;
      for ( p=tablica[hash]; p!=-1; p=nodes[p].next ) {
         int isti = 1, j;
         for ( j=0; j<len; ++j )
            if ( str[ nodes[p].key+j ] != str[ i-len+1+j ] ) {
               isti = 0;
               break;
            }
         if ( isti )
            return 1;
      }

      nodes[nnodes].hash = hash;
      nodes[nnodes].key  = i-len+1;
      nodes[nnodes].next = tablica[hash];
      tablica[hash] = nnodes++;

      hash = (26*hash + sub*str[i-len+1] + str[i+1]) % MAXHASH;
   }

   return 0;
}

int main()
{
   int L, i, lo, hi;
   static char str[MAXLEN+1];
   scanf( "%d%s", &L, str );

   for ( i=0; i<MAXHASH; ++i )
      tablica[i] = -1;

   lo = 0; hi = L-1;
   while ( lo<hi ) {
      int mid = (lo+hi+1) / 2;

      if ( trazi2(str, mid) ) lo = mid;
      else hi = mid-1;
   }

   printf( "%d\n", lo );
   return 0;
}
